doc-content/drools-docs/src/main/asciidoc/DecisionEngine/phreak-sequential-mode-con.adoc

```diff
@@ -0,0 +1,85 @@
+[id='phreak-sequential-mode-con_{context}']
+
+= Sequential mode in Phreak
+
+Sequential mode is a rule base configuration in the {DECISION_ENGINE}, supported by Phreak, that enables rules in a stateless KIE session to access the data modified by the `modify` or `update` statements of previous rules. As a result, rule execution is faster in sequential mode because the data does not change after the initial data set insertion. Rules can be influenced by the execution of other rules for a more intuitive and holistic rule evaluation. Sequential mode also disables inference in rule evaluation.
+
+Sequential mode applies to stateless KIE sessions only because stateful KIE sessions inherently use data from previously invoked KIE sessions. If you use a stateless KIE session but you do not want the execution of previous rules to influence subsequent rules, then ensure that sequential mode is disabled. Sequential mode is disabled by default in the {DECISION_ENGINE}.
+
+To enable sequential mode, use one of the following options:
+
+* Set the system property `drools.sequential` to `true`.
+* Enable sequential mode while creating the KIE base in the client code:
++
+[source,java]
+----
+KieServices services = KieServices.Factory.get();
+KieContainer container = services.newKieContainer(releaseId);
+
+KieBaseConfiguration conf = KieServices.Factory.get().newKieBaseConfiguration();
+conf.setOption(SequentialOption.YES);
+
+KieBase kieBase = kc.newKieBase(conf);
+----
+
+To configure sequential mode to use a dynamic agenda, use one of the following options:
+
+* Set the system property `drools.sequential.agenda` to `dynamic`.
+* Set the sequential agenda option while creating the KIE base in the client code:
++
+[source,java]
+----
+KieServices services = KieServices.Factory.get();
+KieContainer container = services.newKieContainer(releaseId);
+
+KieBaseConfiguration conf = KieServices.Factory.get().newKieBaseConfiguration();
+conf.setOption(SequentialAgendaOption.DYNAMIC);
+
+KieBase kieBase = kc.newKieBase(conf);
+----
+
+When you enable sequential mode, the {DECISION_ENGINE} evaluates rules in the following way:
+
+. Rules are ordered by salience and position in the rule set.
+. An element for each possible rule match is created. The element position indicates the execution order.
+. Node memory is disabled, with the exception of the right-input object memory.
+. The left-input adapter node propagation is disconnected and the object with the node is referenced in a `Command` object. The `Command` object is added to a list in the working memory for later execution.
+. All objects are asserted, and then the list of `Command` objects is checked and executed.
+. All matches that result from executing the list are added to elements based on the sequence number of the rule.
+. The elements that contain matches are executed in a sequence.
++
+NOTE: If you set the maximum number of rule executions, the evaluation network may exit too early.
```

Comment:
In the project docs this wasn't a note but one last point in the list saying: "If we have a maximum number of allowed rule executions, we can exit our network evaluations early to fire all the rules in the elements."

The meaning is quite different. The evaluation may not exit *too* early but can be interrupted for optimization reason without evaluating all the rules. For instance if you set a firing limit of 5, when the 5th activations has been added into the agenda the rule evaluation can be safely interrupted to only fire the 5 activations already found.

Reply:
Ah, got it. How about this then:

```
. The elements that contain matches are executed in a sequence. If you set a maximum number of rule executions, the {DECISION_ENGINE} activates no more than that number of rules in the agenda for execution.
```